### PR TITLE
StreamingDataset: Fault Tolerance v2 2/n

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR refactors the logic to perform fault tolerance. We are now tracking the number of samples yielded from the dataloader and restoring the chunk index and the in chunk index based on the number of yielded samples. 

Fixes #\<issue_number>

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

cc @borda

### CODE REVIEW

File: src/lightning/data/streaming/dataloader.py [hunk 1]
```diff
@@ -370,21 +370,20 @@ def __iter__(self) -> Any:
         else:
             yield from super().__iter__()
 
-    def state_dict(self) -> Optional[Dict[str, Any]]:
```
Comment (awaelchli) [inline src/lightning/data/streaming/dataloader.py:373]: @tchaton We need to follow the Stateful protocol for the state dict. See `lightning.fabric.utilities.types._Stateful`. It would be quite challenging to work with something that is not a dict here. Could we try to make it work?

File: src/lightning/data/streaming/dataset.py [hunk 2]
```diff
@@ -137,9 +125,11 @@ def _create_cache(self, worker_env: _WorkerEnv) -> Cache:
 
     def _create_shuffler(self, cache: Cache) -> Shuffle:
         seed = self.seed
+        drop_last = self.drop_last
         if self._state_dict is not None:
-            seed = self._state_dict[str(cache.rank)]["seed"]
-        return FullShuffle(cache, seed, self.drop_last) if self.shuffle else NoShuffle(cache, seed, self.drop_last)
+            seed = self._state_dict[-1]["seed"]
```
Comment (awaelchli) [inline src/lightning/data/streaming/dataset.py:130]: Can we guarantee that the `self._state_dict` here always contains the key "seed", "drop_last", etc? And why is it correct to choose the one from the last process (-1)?

File: src/lightning/data/streaming/dataset.py [hunk 3]
```diff
@@ -257,73 +289,34 @@ def __next__(self) -> Any:
...
-                    f,
-                )
-
-            # 4. Move the file to its target position
-            shutil.move(tmp_checkpoint_path, os.path.join(self.cache.checkpoint_dir, "checkpoint.json"))
-
-        self.last_time = time()
-
-    def state_dict(self, num_samples_yielded: int = 0, num_workers: int = 0, batch_size: int = 1) -> Dict[str, Any]:
+    def state_dict(self, num_samples_yielded: int, num_workers: int, batch_size: int) -> List[Dict[str, Any]]:
```
Comment (awaelchli) [inline src/lightning/data/streaming/dataset.py:294]: Same here, a dict would be desired here. If you need to store things in a list, we could do:

```py
return {"the_name_of_the_list_thing": [...]} 
```

File: src/lightning/data/streaming/dataset.py [hunk 4]
```diff
@@ -257,73 +289,34 @@ def __next__(self) -> Any:
...
         if _is_in_dataloader_worker():
             raise RuntimeError("The method `state_dict` should only be called in the main process.")
 
-        if self.cache is None:
-            self.worker_env = _WorkerEnv.detect()
-            self.cache = self._create_cache(worker_env=self.worker_env)
-
-        state_dict: Dict[str, Any] = {}
+        state = [
```
Comment (awaelchli) [inline src/lightning/data/streaming/dataset.py:298]: Very odd, a list with one dict entry? See my comment above. Why not just return the dict?

File: src/lightning/data/streaming/dataloader.py (hunk 1 above)
Comment (tchaton) [reply src/lightning/data/streaming/dataloader.py:373]: Yes, for sure. I was just experimenting with an idea. Let me move to dict.

File: src/lightning/data/streaming/dataset.py (hunk 2 above)
Comment (tchaton) [reply src/lightning/data/streaming/dataset.py:130]: `-1` meant this was the information from the last restart. Right now, we don't allow yet for you to change the metadata. In a future PR, I will add support for you to change the world_size, batch_size, num_workers.

File: src/lightning/data/streaming/dataset.py (hunk 4 above)
Comment (tchaton) [reply src/lightning/data/streaming/dataset.py]: removed.

File: src/lightning/data/streaming/dataset.py [hunk 5]
```diff
@@ -203,6 +192,50 @@ def __iter__(self) -> "StreamingDataset":
...
+
+        # re-shuffle the indexes
+        current_indexes = self.shuffler(current_indexes, self.num_chunks, self.current_epoch, self.chunk_index)
+
+        # skip any indexes already consumed
+        current_indexes = current_indexes[indexes[worker_rank] :]
+        self.current_indexes = current_indexes
+
+        # bump the chunk_index
+        self.chunk_index += 1
```
Comment (lantiga) [inline src/lightning/data/streaming/dataset.py:237]: Just a clarification: here the chunk index is bumped because it's pointing to the next one? We could still be at some point of the previous chunk no?

**Review by lantiga — APPROVED** [review]

**Review by lantiga — APPROVED** [review]
Looks great, just a clarifying question

File: src/lightning/data/streaming/dataset.py (hunk 5 above)
Comment (tchaton) [reply src/lightning/data/streaming/dataset.py:237]: Yes, that's correct, I will probably change this in another PR to make it more clear. But right now, we are bumping when generating the indexes from a new chunk: https://github.com/Lightning-AI/pytorch-lightning/blob/master/src/lightning/data/streaming/dataset.py#L240

File: src/lightning/data/streaming/dataset.py [hunk 6]
```diff
@@ -457,3 +405,59 @@ def is_integer(value: str) -> bool:
...
+            worker_intervals.append(chunk_interval)
+
+        workers_chunks[worker_idx] = worker_chunks
+        workers_intervals[worker_idx] = worker_intervals
+
+    return workers_chunks, workers_intervals
+
+
+def _replay_sampling(num_samples_yielded: int, batch_size: int, num_workers: int) -> Dict[int, int]:
+    """This function replays the sampling from the dataloader."""
```
Comment (awaelchli) [inline src/lightning/data/streaming/dataset.py:433]: I don't see any sampling happening in this function. The while loop below can be eliminated completely by setting the values directly, no?

Essentially, 

`indexes[worker_idx] = X * batch_size`, and all you need to compute is X, how many times does worker_idx get to process a batch.

**Review by awaelchli — APPROVED** [review]

File: src/lightning/data/streaming/dataset.py (hunk 6 above)
Comment (tchaton) [reply src/lightning/data/streaming/dataset.py:433]: @awaelchli Thanks, this function was not optimized and would have been expensive with trillion of samples seen, I pushed a tiny update.

This `indexes[worker_idx] = X * batch_size` won't work but this would:

```python
    divisible_num_batches_yielded = num_samples_yielded // (num_workers * batch_size)

    indexes = {}
    for worker_idx in range(num_workers):
        indexes[worker_idx] = divisible_num_batches_yielded * batch_size

    num_samples_yielded = num_samples_yielded - (num_workers * divisible_num_batches_yielded * batch_size)
````

And then, we can take care of the reminder.
